packages/eas-json/src/types.ts

```diff
@@ -25,4 +25,7 @@ export interface EasJson {
   };
   build?: { [profileName: string]: EasJsonBuildProfile };
   submit?: { [profileName: string]: EasJsonSubmitProfile };
+  update?: {
```

Comment:
We should avoid breaking changes in eas.json, so if are deciding on an entirely new config we should approach it carefully.
e.g. maybe putting properties in an update object is fine, but if we will need to have update profiles in the future it might be hard to make that change.
@brentvatne @jonsamp any opinions on that format?

Also if you are adding new properties to eas.json, then https://github.com/expo/eas-cli/blob/main/packages/eas-json/schema/eas.schema.json needs to be updated. That schema is used for code completion and docs in vscode extension.

Reply:
Ah, good call. Yeah, I could definitely imagine that having update configuration be per-profile, though we don't yet have the profile concept in the updates commands. I'm probably not the best person to assess though.

Reply:
should `assetExcludePatterns` be part of the expo updates spec? if so, maybe this config would be better suited for the `updates` key in **app.json**?

Reply:
@brentvatne - probably not since it's really just a (purposeful) limitation of the EAS Update service and is only relevant for uploading assets to EAS Update. The library itself is agnostic to where assets come from. One could imagine an expo-updates server implementation that supports unlimited assets, or one that doesn't serve any assets.

Reply:
got it. while i'm personally not a huge fan of this approach of limiting assets, i understand the practical constraints involved here and why we need it - at least for now.

a couple thoughts:
- should we let people know that this is dangerous? it needs to be factored in to the runtime version of the app and we generally should not recommend using it - it's an escape hatch should people need to temporarily work around the asset limit. eg: `dangerousAssetExcludePatterns`
- wdyt about using a flag instead of an eas.json key? `--dangerously-asset-exclude-patterns="..."`. this would further emphasize that this is not something we would encourage people to depend on.

Reply:
Actually this isn't a temporary solution as it is a bad end-user-experience to have to download many thousands of assets from EAS update. 

Technically EAS update supports unlimited assets (though our uploaded manifest size is limited for the same reason, and the limit is 1mb so referencing that many assets usually puts the manifest over the limit as well). I've been able to upload 9k assets without issue (except manifest limit), ~4k until the manifest limit is hit.

So, we actually will probably want to prescribe it for apps that have more than ~1k assets as then that's offloading a huge amount of data to their users.

> should we let people know that this is dangerous? it needs to be factored in to the runtime version. eg: dangerousAssetExcludePatterns

Most definitely, and that's going to be the hardest part of this task. It might be better to just indicate during the `eas update` command that there are excluded assets (as this PR does), maybe with a more noticeable warning about runtime version?

> wdyt about using a flag instead of an eas.json key? --dangerously-asset-exclude-patterns="...". this would further emphasize that this is not something we would encourage people to depend on.

As mentioned above, this is something that I think we want people to use and depend on.

Reply:
> Actually this isn't a temporary solution as it is a bad end-user-experience to have to download many thousands of assets from EAS update.

I agree that isn't ideal, although I think it's an edge case and it generally would be mitigated in two ways:

1. on the client, we don't download assets if they are already included in the build
2. update configuration includes something like `assetBundlePatterns` (I think this still exists?) that allows you to specify which assets are not part of the atomic update and instead will be downloaded on demand after the update has launched

So the end-user would only ever download thousands of assets in extreme edge cases such as the developer does a build with very few assets, then adds a bunch of assets to their app, and then publishes an update without doing a new build. Is this case more important than I understand it to be or what am I missing?

> Technically EAS update supports unlimited assets (though our uploaded manifest size is limited for the same reason, and the limit is 1mb so referencing that many assets usually puts the manifest over the limit as well). I've been able to upload 9k assets without issue (except manifest limit), ~4k until the manifest limit is hit.

That's great!

Reply:
Adding @ide since he may have more opinions about this.

While I agree it's an extreme edge case, I also don't think we planned on raising the limit, so it seems we may be at some sort of impasse? The manifest size limit helps to ensure that our servers stay sane.

Reply:
I don't want to block this because I think it's important for classic updates sunset. My only request here is that, whatever we end up calling this field or flag, I would like the name to include `dangerous` and/or `experimental`. This indicates very clearly that this is not a well supported path and it gives us more freedom to revisit it in the future, when we have more data in the wild from people who need it.

Reply:
We give the best developers more control when they need it with this feature. A fuller asset specification might let developers specify whether to treat assets as atomic parts of the update (the default behavior, which makes assets available offline, just the same as assets belonging to the embedded update), lazy parts of the update (downloaded when the app needs them like a promotional image banner), or part of the runtime instead of the update (always embedded, never downloaded). Taking this feature all the way, assets that are part of the runtime probably should be excluded from the embedded update as well and be only traditional assets in the app bundle.

This would also help with the ANR problem (app is slow to launch because of the number of asset files to check) as well. It also improves the speed of fetching update manifests since the excluded assets aren't listed in the manifest -- manifests for large apps tend to mostly contain asset-related info.

A separate but correlated issue from the number of assets is the size of assets. Games in particular have large assets like sprites, models, cutscenes, and sometimes many fonts that the developer intends to keep relatively stable. We'd be giving developers the choice to omit uploading those assets when publishing them and checking them when an app downloads an update.

Since this feature needs a clear understanding of runtimes to use successfully, we want to communicate that this feature is advanced and also not introduce this feature until apps could benefit from it. If we were to use the naming of the JSON API to make developers aware of this, a word like "advanced", "expert", or "caution" strikes a balance closer to what we mean IMO. I think "experimental" is also OK if our intent is to write back to developers who we know could benefit from slimming down their updates and let them try this new field. Perhaps "runtime asset" and "update asset" are the right way to think about this rather than "excluded asset" without clarifying "asset excluded from updates and included in the runtime".